src/cfn/template.yaml

```diff
@@ -456,6 +356,340 @@ Resources:
       SubnetId: !Ref PublicSubnet3
       NetworkAclId: !Ref PublicNetworkAcl
 
+  DynamoDBEndpoint:
+    Type: AWS::EC2::VPCEndpoint
+    Properties:
+      RouteTableIds:
+        - !Ref PrivateRouteTable1
+        - !Ref PrivateRouteTable2
+        - !If [ HasThreeAZs, !Ref PrivateRouteTable3, !Ref 'AWS::NoValue' ]
+      ServiceName: !Sub 'com.amazonaws.${AWS::Region}.dynamodb'
+      VpcId: !Ref VPC
+
+  S3Endpoint:
+    Type: AWS::EC2::VPCEndpoint
+    Properties:
+      RouteTableIds:
+        - !Ref PrivateRouteTable1
+        - !Ref PrivateRouteTable2
+        - !If [ HasThreeAZs, !Ref PrivateRouteTable3, !Ref 'AWS::NoValue' ]
+      ServiceName: !Sub 'com.amazonaws.${AWS::Region}.s3'
+      VpcId: !Ref VPC
+
+  # Custom Resources
+
+  LambdaSetup:
+    Type: AWS::Serverless::Function
+    Properties:
+      FunctionName: MediaAnalysisSetup
+      Runtime: nodejs8.10
+      Handler: index.handler
+      CodeUri: ../backend/functions/setup/
+      Description: Custom Lambda resource for the Media Analysis Cloudformation Stack
+      MemorySize: 128
+      Timeout: 30
+      Role: !GetAtt CustomResourceRole.Arn
+
+  CustomResourceRole:
+    Type: AWS::IAM::Role
+    Properties:
+      AssumeRolePolicyDocument:
+        Statement:
+          - Effect: Allow
+            Principal:
+              Service:
+                - lambda.amazonaws.com
+            Action:
+              - sts:AssumeRole
+      Path: /
+      Policies:
+        - PolicyName: MediaAnalysis-setup-S3-fc
+          PolicyDocument:
+            Statement:
+            - Effect: Allow
+              Action: 's3:*'
+              Resource: '*'
+        - PolicyName: MediaAnalysis-cloudwatch-logs
+          PolicyDocument:
+            Statement:
+            - Effect: Allow
+              Action: 'logs:*'
+              Resource: 'arn:aws:logs:*:*:*'
+        - PolicyName: MediaAnalysis-setup-ecr
+          PolicyDocument:
+            Statement:
+            - Effect: Allow
+              Action: 'ecr:DeleteRepository'
+              Resource: '*'
+
+  SetupWebUIAndDockerImage:
+    Type: Custom::Setup
+    DependsOn: CodePipeline
```

Comment:
Turns out that I don't need the custom resource to trigger the pipeline. Given S3 is the trigger, if I just make sure that the pipeline exists before putting the backend.zip to the build bucket, then that will automatically trigger the build and voila. The only issue is that the CFN template may finish the build before the Pipeline completed (in my experience it takes just a couple of minutes), in case, we can do 2 things:
1) We can make the lambda orchestrator to check if the image exists and if not, wait the pipeline to complete in case someone immediately tries to create a fargate instance like 1 second after deploying
2) We can create another lambda custom resource that just looks into the pipeline API and checks if a pipeline is running and keeps waiting until it's done, so that the CFN template end event is blocked by that.

I see that for now, we can go with 1 but I am not super opinionated.

Reply:
Yeah, I think 1 is the simplest for the moment.